### ISSUE TITLE
Questions regarding shar

### ISSUE THREAD

**Issue by marcoyang1998** [lhotse-speech/lhotse#1440]
Hi @pzelasko ,

I am looking at shar format and would greatly appreciate your insights on the following questions:

1. When shall we expect shar to be faster than using a normal CutSet (e.g. lazily loaded from a jsonl.gz)? For example, when using a very large manifest for training (over 30M samples) performing multi-node multi-gpu training?

2. Suppose I have already prepared a manifest (without using the shard) and each cut has various fields like fbank, recording and supervisions. Each cut also has custom arrays (e.g. speaker embedding, codebook indexes), which are stored on disk. What's the best way of creating a shar version for it? 
3.  Do all lhotse samplers support shar? And can I concatenate or use CutSet.mux() to combine multiple cutsets initialized from shar?

4. What shard_size is recommended when constructing shar? Any rules of thumb?

Thank you!

**Comment by marcoyang1998** [comment]
Regarding question 2, a more specific question is: do I need to export the whole manifest again (i.e. exporting all individual fields like fbank features and all customer fields)? Or simply calling `lhotse shar export my_cuts.jsonl.gz out_dir` without specifying fields is enough to benefit from the speed-up of sharding?

**Comment by pzelasko** [comment]
> 1. When shall we expect shar to be faster than using a normal CutSet (e.g. lazily loaded from a jsonl.gz)? For example, when using a very large manifest for training (over 30M samples) performing multi-node multi-gpu training?

Primarily when your storage is slow enough that dataloading becomes a bottleneck for training. Think magnetic disks, NFS, or object stores where individual file access has high overhead. If you have extremely fast SSDs with extremely fast interconnects you can probably get away without it (but if lots of people use the cluster, you will still reduce the overhead about ~5x even for super fast setups).

If you have a multi-node setup, Shar is also a good idea, because the cut samplers can get quite inefficient for large world sizes in the "default" mode.

> 2. Suppose I have already prepared a manifest (without using the shard) and each cut has various fields like fbank, recording and supervisions. Each cut also has custom arrays (e.g. speaker embedding, codebook indexes), which are stored on disk. What's the best way of creating a shar version for it?

Use `lhotse shar export` specifying every field in your data. It will create a copy of each field, but that is necessary in order to perform sequential IO (accelerated data access). If you don't specify the fields, it will still fetch data from individual files (I should have made it drop data instead to avoid confusing people).

> 3. Do all lhotse samplers support shar? And can I concatenate or use CutSet.mux() to combine multiple cutsets initialized from shar?

Yes, every sampler works as usual, so does mux. Follow the Shar tutorial - it is important to use `IterableDatasetWrapper` and `make_worker_init_fn` to set up everything correctly.

> 4. What shard_size is recommended when constructing shar? Any rules of thumb?

Generally aim for 300-1000, lower shard sizes allow more mixing of data across shards in the shuffle/bucket buffers. With lots of GPUs and workers this becomes a bit less relevant.

**Comment by marcoyang1998** [comment]
Thank you so much for the valuable answers @pzelasko ! 

I have a follow up question. Can I skip re-exporting fields that are unnecessary for my training? For example, if I am only using the pre-computed fbank features, can I skip exporting recording? I did this and the code throws warning, but I suppose it's ok?
```/opt/conda/lib/python3.10/site-packages/lhotse/shar/writers/shar.py:144: UserWarning: Found cut with 'recording' field that is not specified for Shar writing. warnings.warn(```

Another question is, when using large training manifests, is it recommended to use `PrecomputedFeatures` or `OnTheFlyFeatures` in terms of speed and efficiency?

**Comment by marcoyang1998** [comment]
I managed to convert my manifest to shar format. However, when I started training, the memory usage of my machine gradually increased and blew up in the end. I suspect that it's loading the fbank features and custom fields to memory. 

My cut (after conversion to shar) looks like this:
```
{"id": "YOU1000000000_S0000050", "start": 0, "duration": 5.325, "channel": 0, "supervisions": [{"id": "YOU1000000000_S0000050", "recording_id": "YOU1000000000_S0000050", "start": 0.0, "duration": 5.325, "channel": 0, "text": "AH AND IT'S SOUND LIKE AS TONS OF POWER AND IT'S FRONT WHEEL DRIVE I AM NOT A FAN OF FRONT WHEEL DRIVE PERFORMANCE CARS YOU KNOW "}], "features": {"type": "kaldi-fbank", "num_frames": 533, "num_features": 128, "frame_shift": 0.01, "sampling_rate": 16000, "start": 0, "duration": 5.325, "storage_type": "shar", "storage_path": "", "storage_key": "", "channels": 0}, "recording": {"id": "YOU1000000000_S0000050", "sources": [{"type": "file", "channels": [0], "source": "download/gigaspeech/data/audio/dev_files/dev_chunks_0000/YOU1000000000_S0000050.wav"}], "sampling_rate": 16000, "num_samples": 85200, "duration": 5.325, "channel_ids": [0]}, "custom": {"codebook_indexes": {"array": {"storage_type": "shar", "storage_path": "", "storage_key": "", "shape": [266, 16]}, "temporal_dim": 0, "frame_shift": 0.02, "start": 0}}, "type": "MonoCut"}
```
My code for loading the shar is as below:
```python
cuts = CutSet.from_shar(
    in_dir="/path/to/shar",
    shuffle_shards=True,
    stateful_shuffle=True,
).repeat()
train_sampler = DynamicBucketingSampler(
    cuts,
    shuffle=True,
    max_duration=100.0,
    num_buckets=10,
    rank=0,
    world_size=1,
)
train_dataset = K2SpeechRecognitionDataset(
    input_transforms=[
        SpecAugment(),  # default configuration is well-tuned
    ],
    input_strategy=PrecomputedFeatures(),
)
train_iter_dataset = IterableDatasetWrapper(
    dataset=train_dataset,
    sampler=train_sampler,
)
train_dloader = DataLoader(
        train_iter_dataset,
        batch_size=None,
        num_workers=2,
        worker_init_fn=make_worker_init_fn(seed=0),
    )

```
After that, loading the first batch takes very long time and the memory usage is growing when calling `for batch in train_dloader`. 

Is loading the features to memory expected? If not, could you please provide some clarification on this? Thanks!

**Comment by pzelasko** [comment]
> I have a follow up question. Can I skip re-exporting fields that are unnecessary for my training? For example, if I am only using the pre-computed fbank features, can I skip exporting recording? I did this and the code throws warning, but I suppose it's ok? `/opt/conda/lib/python3.10/site-packages/lhotse/shar/writers/shar.py:144: UserWarning: Found cut with 'recording' field that is not specified for Shar writing. warnings.warn(`

Yes you can skip the fields you won’t use. 

> Another question is, when using large training manifests, is it recommended to use `PrecomputedFeatures` or `OnTheFlyFeatures` in terms of speed and efficiency?

on the fly feature extraction should be very quick when using lhotse shar to load audio

**Comment by pzelasko** [comment]
> Is loading the features to memory expected? If not, could you please provide some clarification on this? Thanks!

Yes. It has to load the actual data into memory in order to perform shuffling and bucketing on the fly. You can either drop some fields (eg don’t load recordings if using precomputed features) or decrease the shuffle buffer size / buffer size / num workers.

**Comment by marcoyang1998** [comment]
By adjusting buffer size I managed to run everything smoothly. Using Shar is indeed much faster! Thank you and merry Christmas!
